src/libraries/Microsoft.PowerFx.Core/Public/Values/CollectionTableValue.cs

```diff
@@ -19,23 +20,51 @@ namespace Microsoft.PowerFx.Types
     public abstract class CollectionTableValue<T> : TableValue
     {
         // Required - basic enumeration. This is the source object. 
-        private readonly IEnumerable<T> _enumerator; // required. supports enumeration;
+        private IEnumerable<T> _enumerator; // required. supports enumeration;
+
+        // Async implementation
+        private IAsyncEnumerable<T> _asyncEnumerable;        
 
         // Additional capabilities. 
-        private readonly IReadOnlyList<T> _sourceIndex; // maybe null. supports index. 
-        private readonly IReadOnlyCollection<T> _sourceCount; // maybe null. supports count;
-        private readonly ICollection<T> _sourceList; // maybe null. supports mutation;
+        private IReadOnlyList<T> _sourceIndex; // maybe null. supports index. 
```

Comment:
>IReadOnlyList [](http://example.com/codeflow?start=16&length=13)

Can we keep these readonly. 

I see the Init() method, but there are other patterns (like chaining ctors).

Reply:
Except that we can't use different base() calls here, hence this non-optimal logic, I agree

Can be considered for later as this is not the most important [private prop] (in this PR though)